### PR TITLE
Optimize ArrayAllocator CPU usage in OptimizedPartitionedOutputOperator

### PR DESCRIPTION
**Pull request by yingsu00**
The SimpleArrayAllocator takes a substantial CPU amount in OptimizedPartitionedOutputOperator. This PR introduces UncheckedStackArrayAllocator which is 7x faster than SimpleArrayAllocator. The new UncheckedStackArrayAllocator does not keep track of borrowed arrays, therefore the user shall make sure there are no duplicate returnArray calls. 

```
== RELEASE NOTES ==

General Changes
* Remove max buffer count config property `driver.max-page-partitioning-buffer-count` for optimized repartitioning

```

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/operator/UncheckedStackArrayAllocator.java [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+        checkArgument(initialCapacity > 0, "initialCapacity must be positive");
+        intArrays = new int[initialCapacity][];
+        byteArrays = new byte[initialCapacity][];
+    }
+
+    @Override
+    public int[] borrowIntArray(int positionCount)
+    {
+        int[] array;
+        while (intArraysTop >= 0 && intArrays[intArraysTop].length < positionCount) {
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/operator/UncheckedStackArrayAllocator.java:67]: Shouldn't this loop release memory of "discarded" arrays? E.g. intArrays[intArraysTop] = null?

Comment (yingsu00) [reply presto-main/src/main/java/com/facebook/presto/operator/UncheckedStackArrayAllocator.java:67]: @mbasmanova hi Masha, the arrays will be released when the stack grows back, but you're right, we should explicitly release them. I'll make the change in next iteration.

**Comment by yingsu00** [comment]
> @yingsu00 I just skimmed through the diff and have a question below. Also, would you describe the query-wide impact of this change?

Hi Masha, in T67972617, the operator CPU reduced from 660 to about 100 days with this fix, but it's still much slower than the original 2.21 days. The subsequent PRs will have additional fixes. The regressed query, however, uses 3x less memory than before.

**Comment by yingsu00** [comment]
@mbasmanova Masha, I updated the PR with your comments addressed, also updated the TestBlockEncodingBuffers to use the new allocator. Thank you for reviewing!

**Pull request merged**